File: workshop/apps/python/sample/AIWorkshopAgentDevPython/src/graph.py
import httpx
from typing import Optional, List, Any


class GraphService:

    def __init__(self, access_token: str):
        self.access_token = access_token

    async def get_copilot_data(self, query: str, language: Optional[str] = None) -> List[Any]:

        request_body = {
            "queryString": query,
            "dataSource": "sharePoint",
            "resourceMetadata": ["title", "url"],
            "maximumNumberOfResults": 5,
        }

        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
            "SdkVersion": "ContosoEnterpriseSearchAgent",
            "Accept-Language": language if language else "en-US",
        }

        url = "https://graph.microsoft.com/v1.0/copilot/retrieval"

        async with httpx.AsyncClient() as client:
            response = await client.post(url, json=request_body, headers=headers)

            if response.status_code == 200:
                data = response.json()
                return data.get("retrievalHits", [])

            print(f"Error: {response.status_code} - {response.text}")
            return []

File: workshop/apps/typescript/sample/AIWorkshopAgentDevNode/src/graph.ts
import { Client } from "@microsoft/microsoft-graph-client";

export class GraphService {

    private graphClient: Client;

    constructor(accessToken: string) {

        this.getCopilotData = this.getCopilotData.bind(this);
        this.graphClient =  Client.initWithMiddleware(
            {
                authProvider: {
                    getAccessToken: async () => {
                        return accessToken
                    },
                }
            } 
        );
    }      

    public async getCopilotData(query: string, language?: string): Promise<any[]> {
    
        const response = await this.graphClient.api("/copilot/retrieval").headers({ 
              "SdkVersion": "ContosoEnterpriseSearchAgent",  
              "Accept-language": language ? language : "en-US"
            }).post({
            queryString: query,
            dataSource: "sharePoint",
            resourceMetadata: ['title','url'],
            maximumNumberOfResults: 5
        });

        return response.retrievalHits;
    }
}